### PR TITLE
Allow specifying vcpkg tool path in X_VCPKG_TOOL_PATH

### LINKED ISSUE #31641: There should be a better way to download vcpkg-tool for offline development environments
I understand that vcpkg itself is not really designed for offline development environments because it downloads packages from the internet, but if necessary it can be made to work in an offline development environment by setting up the binary cache in a remote source that the offline development environment has access to, and then push to that remote source from an internet-connected machine.

However, there is one problem with this approach, which is that the offline development environment still needs to access the internet in order to download `vcpkg-tool` during bootstrapping. Currently the location of `vcpkg-tool` is [hardcoded](https://github.com/microsoft/vcpkg/blob/c6592ce60ff394049905365865f59e3a4d93d35b/scripts/bootstrap.sh#L197) in the bootstrap script and is not configurable in any way as far as I can tell.

I can think of the following potential solutions to resolve this issue:

1. Clone the `vcpkg` and `vcpkg-tool` repositories to an internal git server and modify `bootstrap.sh` to point to that address instead of github.com.
2. Add back a way to build the `vcpkg` executable from source, like there used to be a few years ago.
3. Add a new environment variable called `VCPKG_TOOL_URL` or `VCPKG_TOOL_PATH` or something like that to tell `bootstrap.sh` where to pull `vcpkg-tool` from.
4. Somehow upload the `vcpkg` executable itself to the remote source as part of the binary caching step, and retrieve from the CMake toolchain file as the first step of the offline bootstrapping process.
5. Something else?

Which is these is the best solution to the problem, and what would be the timeline for implementing it?

### PR DESCRIPTION
**Pull request by sweemer**
Fixes #31641. Please see that issue for all of the details of the problem I am trying to solve.

I am not an expert in powershell, so it would be good if someone could review that part especially closely for me.

Also please let me know the best place to add some documentation on the `VCPKG_TOOL_PATH` environment variable, and I will do so in a separate pull request.

N.B. please view the diffs without whitespace (add `&w=1` to the URL) to make it easier to read.

### CODE REVIEW

**Review by Cheney-W — DISMISSED** [review]

**Comment by Cheney-W** [comment]
@sweemer About documentation, you could submit issue or pr at https://github.com/microsoft/vcpkg-docs.

**Comment by Cheney-W** [comment]
@sweemer Could you please apply vicroms's suggestion and solve the conflicts? Thank you!

File: scripts/bootstrap.ps1 [hunk 1]
```diff
@@ -45,22 +45,37 @@ while (!($vcpkgRootDir -eq "") -and !(Test-Path "$vcpkgRootDir\.vcpkg-root"))
 
 Write-Verbose "Examining $vcpkgRootDir for .vcpkg-root - Found"
 
-# Read the vcpkg-tool config file to determine what release to download
-$Config = ConvertFrom-StringData (Get-Content "$PSScriptRoot\vcpkg-tool-metadata.txt" -Raw)
-$versionDate = $Config.VCPKG_TOOL_RELEASE_TAG
+if ($env:VCPKG_TOOL_PATH -eq $null)
```
[the remark this answers is not in the document]
Comment (sweemer) [reply scripts/bootstrap.ps1]: Done

**Comment by BillyONeal** [comment]
I'm not convinced of this for the same reasons I said back in https://github.com/microsoft/vcpkg/issues/31641#issuecomment-1569283248 . What is the intended use case and how is the customer expected to use this feature? Why is:

```
:: your custom code that gets vcpkg on the airgapped machine
set X_VCPKG_TOOL_PATH=blah
cmake -G ninja -DCMAKE_TOOLCHAIN_FILE=%VCPKG_ROOT%\scripts\buildsystems\vcpkg.cmake
```

any easier than

```
:: your custom code that gets vcpkg on the airgapped machine
copy /Y blah\vcpkg.exe %VCPKG_ROOT%\vcpkg.exe
cmake -G ninja -DCMAKE_TOOLCHAIN_FILE=%VCPKG_ROOT%\scripts\buildsystems\vcpkg.cmake
```

or

```
:: your custom code that gets vcpkg on the airgapped machine
set X_VCPKG_TOOL_PATH=blah
.\bootstrap_vcpkg.bat
```

any easier than

```
:: your custom code that gets vcpkg on the airgapped machine
copy /Y blah\vcpkg.exe %VCPKG_ROOT%\vcpkg.exe
```

?

Of particular concern of adding such a feature is that it would 'sell' that mixing and matching `vcpkg.exe`s and the 'scripts' directory is supported and intended, which is not the case. (New tools take dependencies on changes to scripts tree behavior fairly frequently)

**Comment by sweemer** [comment]
Hi @BillyONeal thanks for the review, and here are answers to your questions:

> What is the intended use case 

To bootstrap vcpkg offline through the CMake integration without any manual actions required. 

> how is the customer expected to use this feature? 

By setting the value of the `X_VCPKG_TOOL_PATH` environment variable e.g. in `CMakePresets.json`.

> Why is [it] any easier?

Because it eliminates the need for manual actions before bootstrapping vcpkg. Even something as simple as copying a file from one location to another is vulnerable to human error, and adds unnecessary complexity to the development process.

> Of particular concern of adding such a feature is that it would 'sell' that mixing and matching vcpkg.exes and the 'scripts' directory is supported and intended, which is not the case.

Isn't it more concerning that developers are forced to remember when to manually copy over a new vcpkg.exe in their air-gapped environment? I strongly believe that it's better to manage the mapping in source control instead.

For example, what I'm planning to do is put vcpkg.exe in a versioned location such as `/internal-network-drive/vcpkg-tool-YYYY-MM-DD/vcpkg.exe` and then set `X_VCPKG_TOOL_PATH=/internal-network-drive/vcpkg-tool-YYYY-MM-DD` in my `CMakePresets.json` file. Then any developer in my team can clone our project in our internal network and immediately bootstrap vcpkg (a submodule of our project) with the correct version of vcpkg.exe. Different developers can be on different branches with different versions of the vcpkg submodule and all be confident that they are using the correct version of vcpkg.exe at all times.

**Comment by vicroms** [comment]
You could get the desired behavior today without modifying the bootstrap script by adding these lines before the first call to `project()` in your `CMakeLists.txt` files.

```cmake
if (NOT EXISTS "$ENV{VCPKG_ROOT}/vcpkg${CMAKE_EXECUTABLE_SUFFIX}")
  file(COPY "${X_VCPKG_TOOL_PATH}" DESTINATION "$ENV{VCPKG_ROOT}")
endif()
```

Then you can just set `X_VCPKG_TOOL_PATH` (or any name you want to give it) in `CMakePresets.json`.

**Comment by sweemer** [comment]
@vicroms Wow, that's great! I didn't know that the toolchain file was executed during the call to `project()` - I thought it was executed before entering `CMakeLists.txt`, which is why I kept trying to add this functionality through the toolchain file.

I think it would be very helpful to document this somewhere. I will close this pull request and add your snippet to https://github.com/microsoft/vcpkg-docs/pull/279 if you don't mind.

**Pull request closed without merging**